### DISCUSSION TITLE
clarification about ripgrep functionality and possible example

### DISCUSSION THREAD

**Discussion by atereou** [BurntSushi/ripgrep#1579] [General] [answered]
Ripgrep is capable of searching any file content for text and can also process a whole drive in theory?
I don't know if this works, but if it does - how do I solve this in Windows exactly?
I'm in this folder C:\Users\Administrator\search\rg.exe
and my test file is here C:\Users\Administrator\search\ui\test.exe
test.exe contains the string below and ripgrep must find it:
E X I T   H E L P
(**E**_space_**X**_space_**I**_space_**T**_spacespacespace_**H**_space_**E**_space_**L**_space_**P**)
When I now run CMD in the 'search' folder, is there a command for ripgrep that will process the whole 'search' folder with subfolders?
How would I write the command exactly to find E X I T   H E L P in all files with .exe extension only?
Or is this totally not what ripgrep can do?

**Comment by BurntSushi** [comment, marked as answer]
> When I now run CMD in the 'search' folder, is there a command for ripgrep that will process the whole 'search' folder with subfolders?

You should just be able to run `.\rg.exe "E X I T  H E L P"` and it will recursively search your current directory. If you only want to search exe files, then you can add `-g "*.exe"`. Presumably, exe files are in a binary format, so you'll probably also need to force ripgrep to treat it as text, which means using the `-a` (short for `--text`). So I think your complete command is probably:

```
$ .\rg.exe -g "*.exe" -a "E X I T  H E L P"
```

Note though that I am not a Windows user.

**Reply by atereou** [reply]
Yes, this works. Thanks!
Then I have one final question about the topic.
If the file is not EXE binary, but something like *.psd from Adobe Photoshop or Gimp *.xcf project files - are these file formats also supported? In this case what do I have to change in the command?
I mean can I specify here *.psd as search filter too or can I only do below in this case?
rg -a "texttofindinpsdfile" "path"

**Reply by BurntSushi** [reply]
ripgrep doesn't know about file formats. To ripgrep, files are just bytes. It reads and searches those bytes as is. The `-a` flag is necessary when searching binary formats because ripgrep will otherwise try to skip them, since you _usually_ don't want to search binary files with a tool like ripgrep.

But sure, you can use `-g "*.psd"` if you want. But you don't need `-g` for it. `-g` just limits your search to file paths that match the glob. Otherwise, ripgrep will try to search everything.

**Reply by atereou** [reply]
Amazing - this works too. Thanks a lot for your answers!
I was going through 100 pages search results on Github to find a good search tool for Windows that can do all of this and which is eventually faster than the tool TextCrawler which beated all other tools I tested so far, but was still a bit slow for my requirements.
After doing some more testing I can confirm that ripgrep is blazingly fast in comparison.

**Reply by BurntSushi** [reply]
Great, glad to hear it!
